agentGroupPresence/agentGroupPresenceController.js

```diff
@@ -0,0 +1,147 @@
+        var matchingRow = $(`#${a.id}`);
+        if(matchingRow.length == 0) {
+            matchingRow = addRow(a);
+        }
+
+        var query = $(matchingRow);
+        query.find('div.Name')
+            .text(a.name);
+
+        a.statuses.forEach(s => {
+            if(this.agentIds.indexOf(s.id) === -1){
+                this.agentIds.push(s.id);
+            }
+
+            query.find('div.' + s.type)
+                .text(s.status);
+        });
+    });
+}
+
+function join(button){
+    joinLeave(button, 'Present');
+}
+
+function leave(button){
+    joinLeave(button, 'NotPresent');
+}
+
+function joinLeave(button, status) {
+    var id = button.parentElement.parentElement.id;
+    var type = $(button).siblings('div')[0].className;
+
+    var employeeAgentGroupModel = {};
+    employeeAgentGroupModel.id = id;
+
+    var agentGroupClone = jQuery.extend(true, {}, this.agentGroups);
+    agentGroupClone._embedded.items.forEach(a => {
+        if(a.id === id){
+            employeeAgentGroupModel.name = a.name;
+            employeeAgentGroupModel.reporting = a.reporting;
+            a.statuses.forEach(s => {
+                if(s.type === type){
+                    s.status = status;
+                }
+            });
+            employeeAgentGroupModel.statuses = a.statuses;
+        }
+    });
+    micc.putAgentGroupPresenceByGroupId(`me`, id, JSON.stringify(employeeAgentGroupModel), function processResponse(responseData) {
+        console.log(`Response for agent group presence for ${id} for type ${type} with status ${status}:  `, responseData);
+    });
+}
+
+function joinLeaveAll(button, status) {
+    var id = button.parentElement.parentElement.id;
+    var agentGroupClone = jQuery.extend(true, {}, this.agentGroups);
+    agentGroupClone._embedded.items.forEach(a => {
+        if(a.id === id){
+            a.statuses.forEach(s => {
+                s.status = status;
+            });
+        }
+    });
+    micc.putAgentGroupPresence(`me`, JSON.stringify(agentGroupClone), function processResponse(responseData) {
+        console.log(`Response for all agent group presence for ${id} with status ${status}:  `, responseData);
+    });
+}
+
+function joinAll(button) {
+    joinLeaveAll(button, 'Present');
+}
+
+function leaveAll(button) {
+    joinLeaveAll(button, 'NotPresent');
+}
+
+function addRow(agentGroup) {
+    return $(`<tr id="${agentGroup.id}" class="conversationRow">
+    <td class="JoinAll"><button class="group-button w3-btn-block w3-blue w3-round" onclick="joinAll(this);">Join All</button></td>
+    <td class="LeaveAll"><button class="group-button w3-btn-block w3-blue w3-round" onclick="leaveAll(this);">Leave All</button></td>
+    <td><div class="Name">?</div></td>
+    <td><div class="Voice">N/A</div><button class="individual-button w3-btn-block w3-blue w3-round" onclick="join(this);">Join</button><button class="individual-button w3-btn-block w3-blue w3-round" onclick="leave(this);">Leave</button></td>
```

Comment:
This repeated code can be cleaned up nicely with a create button func that takes the text and a click handler.